lib/vtls/openssl.c

```diff
@@ -1878,15 +1878,45 @@ static void ossl_close(struct Curl_cfilter *cf, struct Curl_easy *data)
 
   if(backend->handle) {
     if(cf->next && cf->next->connected) {
-      char buf[32];
+      char buf[1024];
+      int nread, err;
+      long sslerr;
+
       /* Maybe the server has already sent a close notify alert.
          Read it to avoid an RST on the TCP connection. */
       (void)SSL_read(backend->handle, buf, (int)sizeof(buf));
-
-      (void)SSL_shutdown(backend->handle);
-
       ERR_clear_error();
+      if(SSL_shutdown(backend->handle)) {
+        CURL_TRC_CF(data, cf, "SSL shutdown finished");
+      }
+      else {
+        nread = SSL_read(backend->handle, buf, (int)sizeof(buf));
+        err = SSL_get_error(backend->handle, nread);
+        switch(err) {
+        case SSL_ERROR_NONE: /* this is not an error */
+        case SSL_ERROR_ZERO_RETURN: /* no more data */
+          CURL_TRC_CF(data, cf, "SSL shutdown, EOF from server");
+          break;
+        case SSL_ERROR_WANT_READ:
+          /* SSL has send its notify and now wants to read the reply
+           * from the server. We are not really interested in that. */
+          CURL_TRC_CF(data, cf, "SSL shutdown, sent, leaving");
+          break;
+        case SSL_ERROR_WANT_WRITE:
+          CURL_TRC_CF(data, cf, "SSL shutdown, blocked send, leaving");
+          break;
+        default:
+          sslerr = ERR_get_error();
+          CURL_TRC_CF(data, cf, "SSL shutdown, error: '%s', errno %d",
+                      (sslerr ?
+                        ossl_strerror(sslerr, buf, sizeof(buf)) :
+                        SSL_ERROR_to_str(err)),
```

Comment:
```suggestion
                       ossl_strerror(sslerr, buf, sizeof(buf)) :
                       SSL_ERROR_to_str(err)),
```
also why did you assign sslerr to ERR_get_error and err to SSL_get_error, wouldn't it make more sense to use err for ERR_get_error and sslerr for SSL_get_error?

Reply:
Will fix the indentation. 

As to the assigns: `SSL_get_error()` returns `int`, while `ERR_get_error()` returns `long`.

Reply:
> As to the assigns: `SSL_get_error()` returns `int`, while `ERR_get_error()` returns `long`.

yes I understand but what I'm saying is sslerr makes more sense representing SSL_get_error, so what I mean is wouldn't this make more sense:

long err;
int sslerr;
...
sslerr = SSL_get_error(backend->handle, nread);
...
err = ERR_get_error();

Also wouldn't they be retrieving the same error at this point?

Reply:
I admit to copying this from other places in our code. If you would like to go over the TLS implementations and correct the naming of variables, please do.

As to your question "is this the same"? As far as my memory goes, the `long` is partially the same, but has in addition the OpenSSL component (CRYPTO, etc.) coded into the value. So it includes the `int` in a way that is internal to the OpenSSL implementation, afaik.

Reply:
> I admit to copying this from other places in our code.

Ok I see what you mean, other functions like ossl_shutdown use that naming as well. nevermind